internal/kubernetes/apis.go

```diff
@@ -0,0 +1,58 @@
+// Copyright 2024 Crunchy Data Solutions, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package kubernetes
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// API is a combination of Group, Version, and Kind that can be used to check
+// what is available in the Kubernetes API. There are four ways to populate it:
+//  1. Group without Version nor Kind means any resource in that Group.
+//  2. Group with Version but no Kind means any resource in that GV.
+//  3. Group with Kind but no Version means that Kind in any Version of the Group.
+//  4. Group with Version and Kind means that exact GVK.
+type API = schema.GroupVersionKind
+
+type APIs interface {
+	Has(API) bool
+	HasAll(...API) bool
+	HasAny(...API) bool
+}
+
+// APISet implements [APIs] using empty struct for minimal memory consumption.
+type APISet = sets.Set[API]
+
+func NewAPISet(api ...API) APISet {
+	s := sets.New[API](api...)
+
+	for i := range api {
+		s.Insert(
+			API{Group: api[i].Group},
+			API{Group: api[i].Group, Version: api[i].Version},
+			API{Group: api[i].Group, Kind: api[i].Kind},
```

Comment:
We don't want to insert the full GVK?

Reply:
That's happening on L32. The full extent of what they passed us is added there, and these add the... lesser combos. Worth a code comment, but I'm not sure how to describe it atm.

Reply:
😆 That wasn't so bad.